### PR TITLE
autodedup configuration with final stage

### PR DESCRIPTION
**Pull request by Ayushmina-20**
- Added auto-dedup configuration with final stage
@avikganguly01 @truthfool

### CODE REVIEW

File: openg2p_registration/models/openg2p_create_registration.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+from odoo.tools.translate import _
+import time
+
+AVAILABLE_PRIORITIES = [("0", "Urgent"), ("1", "High"), ("2", "Normal"), ("3", "Low")]
+
+
+class CreateRegistration(models.Model):
+    _inherit = ["openg2p.registration"]
+
+    def create_registration_from_odk(self, odk_data):
```
Comment (avikganguly01) [inline openg2p_registration/models/openg2p_create_registration.py:12]: That's not how functions should be defined. Usw function overriding.

File: openg2p_registration/models/openg2p_create_registration.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+            else:
+                self.refactor_create_fields(temp)
+        elif int(temp["autodedup_id"]) in range(4, 7):
+            externaldup_list = (
+                self.env["openg2p.registration"]
+                .search([("external_id", "=", temp["new_emis_code"])])
+                .ids
+            )
+            if len(externaldup_list) != 0:
+                if temp["autodedup_id"] == "4":
```
Comment (avikganguly01) [inline openg2p_registration/models/openg2p_create_registration.py:49]: Ids shouldn't be used like this. Use enums instead.

File: openg2p_registration/models/openg2p_create_registration.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+        state_name = temp["state"] if "state" in temp.keys() else "Freetown"
+        country_id = self.env["res.country"].search([("name", "=", country_name)])[0].id
+        state_id = (
+            self.env["res.country.state"].search([("name", "=", state_name)])[0].id
+        )
+        # Creating registration in final stage
+        stage_id = 6
+        temp["stage_id"] = int(temp["next_stage"])
+        try:
+            regd = self.create(
```
Comment (avikganguly01) [inline openg2p_registration/models/openg2p_create_registration.py:80]: Please explain why this new piece of creating a registration is added with such values.

**Review by avikganguly01 — CHANGES_REQUESTED** [review]

File: odk-connector/models/odk_config_model.py [hunk 4]
```diff
@@ -52,6 +52,32 @@ class ODKConfig(models.Model):
...
+            ("3", "Next Field:KYC Action:Delete new data"),
+            ("4", "Next Field:External id Action:merge"),
+            ("5", "Next Field:External id Action:Delete old data"),
+            ("6", "Next Field:External id Action:Delete new data"),
+        ],
+        string="Autodedup Configuration",
+        required=True,
+    )
+
+    next_stage = fields.Selection(
```
[the remark this answers is not in the document]
Comment (truthfool) [reply odk-connector/models/odk_config_model.py:69]: Stages are not defined as enums. They are defined in a data module as XML file.

File: odk-connector/models/odk_config_model.py [hunk 5]
```diff
@@ -52,6 +52,32 @@ class ODKConfig(models.Model):
         store=True,
         required=True,
     )
+
+    autodedup_id = fields.Selection(
+        [
+            ("1", "Next Field:KYC Action:Merge"),
```
[the remark this answers is not in the document]
Comment (truthfool) [reply odk-connector/models/odk_config_model.py:58]: Separate the next field(make it as autodedup_field) and actions and add the complete configuration in the openg2p program.

File: odk-connector/models/odk_config_model.py [hunk 6]
```diff
@@ -52,6 +52,32 @@ class ODKConfig(models.Model):
...
+            ("5", "Next Field:External id Action:Delete old data"),
+            ("6", "Next Field:External id Action:Delete new data"),
+        ],
+        string="Autodedup Configuration",
+        required=True,
+    )
+
+    next_stage = fields.Selection(
+        [
+            ("1", "New Stage"),
```
[the remark this answers is not in the document]
Comment (truthfool) [reply odk-connector/models/odk_config_model.py:71]: Reuse the reference for stages from openg2p.registration.stage. Do not hardcode the stages and their ids.

File: openg2p_registration/models/openg2p_create_registration.py (hunk 1 above)
Comment (truthfool) [reply openg2p_registration/models/openg2p_create_registration.py:12]: Function names should not be defined for function parameters or where the function is being called from.

File: openg2p_registration/models/openg2p_registration.py [hunk 7]
```diff
@@ -680,6 +681,9 @@ def create_beneficiary_from_registration(self):
             "emergency_contact": self.emergency_contact,
             "emergency_phone": self.emergency_phone,
             "odk_batch_id": self.odk_batch_id,
+            "kyc_id": self.kyc_id,
```
Comment (avikganguly01) [inline openg2p_registration/models/openg2p_registration.py:684]: Why add this as new field? It already exists as part of registration model linked with KYCs.

File: openg2p_registration/models/registration_service.py [hunk 8]
```diff
@@ -9,10 +10,77 @@ class RegistrationService(models.Model):
...
 
-    def create_fields_for_registration(self, temp):
+            if len(kycdup_list) != 0:
+                if action == "merge":
+                    # merging the existing ones with new one and passing the existing_id and current_id as arguments
+                    self.merge_registrations(temp, kycdup_list[0])
+                elif action == "del_old":
+                    # deleting the old registration
+                    duplicate_kyc = self.env["openg2p.registration"].search(
+                        [("kyc_id", "=", temp["bban"])]
```
Comment (avikganguly01) [inline openg2p_registration/models/registration_service.py:38]: Why are we searching kyc ID with BBAN field?

File: openg2p_registration/models/registration_service.py [hunk 9]
```diff
@@ -9,10 +10,77 @@ class RegistrationService(models.Model):
...
+                    self.merge_registrations(temp, externaldup_list[0])
+                elif action == "del_old":
+                    # deleting the old registration
+                    duplicate_external = self.env["openg2p.registration"].search(
+                        [("external_id", "=", temp["new_emis_code"])]
+                    )
+                    duplicate_external.active = False
+                    # deleting the old beneficiary
+                    duplicate_external_bene = self.env["openg2p.beneficiary"].search(
+                        [("external_id", "=", temp["new_emis_code"])]
```
Comment (avikganguly01) [inline openg2p_registration/models/registration_service.py:71]: EMIS code has nothing to do with autodedup other than the part that it's part of ODK submission and is used to populate external ID field of registration.

File: openg2p_registration/models/registration_service.py [hunk 10]
```diff
@@ -9,10 +10,77 @@ class RegistrationService(models.Model):
...
+                    )
+                    duplicate_external_bene.active = False
+                    # creating registration for new record
+                    self.create_fields_for_registration(temp, stage_name)
```
Comment (avikganguly01) [inline openg2p_registration/models/registration_service.py:75]: Why would a function for creating a registration be named create fields for registration? Please change.

File: openg2p_registration/models/registration_service.py [hunk 11]
```diff
@@ -141,6 +209,12 @@ def create_disbursement_fields(self, rid, temp, regd):
                         data["bank_account_id"] = res.id
                 elif k == "phone":
                     data["phone"] = odk_data["phone"]
+                elif k == "bban":
+                    data["kyc_id"] = odk_data["bban"]
+                elif k == "new_emis_code":
+                    data["external_id"] = odk_data["new_emis_code"]
+                elif k == "town_village":
+                    data["town_village"] = odk_data["town_village"]
```
Comment (avikganguly01) [inline openg2p_registration/models/registration_service.py:217]: What does town village have to do with auto deduplication?

**Review by avikganguly01 — CHANGES_REQUESTED** [review]

**Comment by avikganguly01** [comment]
Don't force push as it's making it harder to track the code changes related to the changes requested.

File: openg2p_registration/models/registration_service.py (hunk 9 above)
Comment (Ayushmina-20) [reply openg2p_registration/models/registration_service.py:71]: I am checking duplicates before registration is created so for this reason I am using that field. Here temp is a dictionary containing the odk submissions

File: openg2p_registration/models/registration_service.py (hunk 8 above)
Comment (Ayushmina-20) [reply openg2p_registration/models/registration_service.py:38]: Already data got collected in odk central as "bban".

File: openg2p_registration/models/registration_service.py (hunk 11 above)
Comment (Ayushmina-20) [reply openg2p_registration/models/registration_service.py:217]: Its optional I will remove it.

File: openg2p_registration/models/registration_service.py (hunk 10 above)
Comment (Ayushmina-20) [reply openg2p_registration/models/registration_service.py:75]: How about mapping_fields_for_registration ?

**Comment by Ayushmina-20** [comment]
- Final commit [refactoring and moving KYC to registration identity module]
- Creating a mapping function in odk submission model for core fields and MBSSE fields such as emis code 
- making KYC a part of the registration identity model 
- breaking a large if else if ladder into small functions
- mapping of other fields which required registration id in the registration model
- TODO correct submission field to be mapped for kyc_id
@avikganguly01 @truthfool

**Comment by avikganguly01** [comment]
@Ayushmina-20 : Please rebase with upstream/master to resolve conflicts.

**Comment by Ayushmina-20** [comment]
> @Ayushmina-20 : Please rebase with upstream/master to resolve confIct
I have to force push because normal push is not working

**Pull request merged**